silkworm/infra/grpc/server/server.hpp

```diff
@@ -78,10 +80,17 @@ class Server {
         register_async_services(builder);
 
         server_ = builder.BuildAndStart();
-        SILK_TRACE << "Server " << this << " bound at selected port: " << selected_port;
+
         if (server_ == nullptr) {
+            std::string error_msg = "cannot start gRPC server at " + settings_.address_uri;
+
+            if (ServerGlobalCallbacks::bad_port_error.load(std::memory_order_acquire)) {
+                error_msg += " (port already in use)";
+                ServerGlobalCallbacks::bad_port_error.store(false, std::memory_order_release);
```

Comment:
```suggestion
            if (ServerGlobalCallbacks::bad_port_error) {
                error_msg += " (port already in use)";
                ServerGlobalCallbacks::bad_port_error = false;
```

nit: normally we only use explicit memory order if it matters for performance. Here it doesn't matter, so a default memory order is fine. In this case it is possible to use atomic_bool as a plain bool with a help of conversion and assignment operators that atomic_bool has.

An extra improvement idea is to have a method on ServerGlobalCallbacks to encapsulate read and reset of this flag.

Reply:
Gotcha, seems like a good spot for .exchange() then!